### DISCUSSION TITLE
I2C receive data

### DISCUSSION THREAD

**Discussion by Prandov** [kurbatov/firmata4j#58] [Q&A]
I successfully tested I2CExample.java!
---
Now I'm trying to receive data from an I2C device (SGP30 CO2 sensor)
I created a wrapper for my I2C device as suggested in readme.md 

To receive data from an I2C device: device.ask() requires a listener receive the data. 
I'm struggling to implement this listener, 
Is there an example using `device.ask()` ? 

Thanks

**Comment by kurbatov** [comment]
Please, find the example of an I2C wrapper in this file https://github.com/kurbatov/firmata4j/blob/master/src/main/java/org/firmata4j/ssd1306/SSD1306.java
You don't have to implement `I2CDevce` interface as the library already does so in `FirmataI2CDevice`.

**Reply by Prandov** [reply]
Many thanks for the reply!

So _SSD1306.java_ uses `device.tell` from _FirmataI2CDevice.java_ (implementing the _I2CListener.java_ interface)?
>From _SSD1306.java_ 
227  `device.tell(Arrays.copyOfRange(command, i, i + 2));`

I don't have to use `implements I2CListener` at the top of my class since it is implemented in  _FirmataI2CDevice.java_ ?
Then I'm unsure how to assign the listener to receive the reply: 
>From _FirmataI2CDevice.java_ :
77 `public void ask(byte responseLength, I2CListener listener) throws IOException {`
and
82 `public void ask(int register, byte responseLength, I2CListener listener) throws IOException {`

Is the listener created by the library or do I have to create it?

**Reply by kurbatov** [reply]
Using `ask` will do if you want to receive data from the device only once. In case you want react to a continuous stream of events from the device, there is `subscribe` method that takes a listener which you have to implement writing handling logic in there.
